### ISSUE TITLE
All docker API requests hang

### ISSUE THREAD

**Issue by csams** [docker/docker#19223]
### Description
All requests to the daemon via the docker command hang if managing the daemon via the init.d script.  I think this might be related to #13885, except I'm seeing the hang immediately with zero load/no other requests.

### docker version
```
$ sudo docker version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 06:20:54 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 06:20:54 UTC 2015
 OS/Arch:      linux/amd64
```
### docker info
```
$ sudo docker info
Containers: 0
Images: 0
Storage Driver: devicemapper
 Pool Name: docker-253:2-3154434-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.821 GB
 Data Space Total: 107.4 GB
 Data Space Available: 6.594 GB
 Metadata Space Used: 1.479 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.85 (2014-04-10)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.19.8-100.fc20.x86_64
Operating System: Fedora 20 (Heisenbug)
CPUs: 8
Total Memory: 15.37 GiB
Name: localhost.localdomain
ID: EPD6:QKSE:OD6C:REJA:7FHB:AKHV:IUF7:DDCX:4SO2:GX5U:DQVP:ZKZN
```
### System Info
I'm executing everything locally.
using yum repo at https://yum.dockerproject.org/repo/main/fedora/20/

`$ sudo yum install docker-engine`
installs docker-engine-1.8.3-1.fc20.x86_64

```
$ uname -a
Linux localhost.localdomain 3.19.8-100.fc20.x86_64 #1 SMP Tue May 12 17:08:50 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```
### How Reproducible?
At will.

### Steps to Reproduce
In one terminal execute `$ sudo systemctl start docker` or `sudo service docker start`
In another execute `sudo docker version`

### Actual Results
```
$ sudo strace -s 200 docker version
...
getsockname(5, {sa_family=AF_LOCAL, NULL}, [2]) = 0
getpeername(5, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, [23]) = 0
read(5, 0xc2080c5000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
write(5, "GET /v1.20/version HTTP/1.1\r\nHost: /var/run/docker.sock\r\nUser-Agent: Docker-Client/1.8.3 (linux)\r\n\r\n", 100) = 100
epoll_wait(4, {{EPOLLOUT, {u32=1929426144, u64=139661380990176}}}, 128, 0) = 1
epoll_wait(4, 
```
simply hangs with no further output.

After editing `/etc/init.d/docker` to use strace, the last bit of `/var/log/docker` is:
```
...
munmap(0x7fcd2506f000, 4096)            = 0
futex(0x1721420, FUTEX_WAKE, 1)         = 1
futex(0x17213a0, FUTEX_WAKE, 1)         = 1
getpid()                                = 7124
getuid()                                = 0
getgid()                                = 0
sendmsg(6, {msg_name(0)=NULL, msg_iov(1)=[{"\0", 1}], msg_controllen=32, [{cmsg_len=28, cmsg_level=SOL_SOCKET, cmsg_type=SCM_CREDENTIALS, {pid=7124, uid=0, gid=0}}], msg_flags=0}, 0) = 1
write(6, "AUTH\r\n", 6)                 = 6
read(6, 0xc208197000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {}, 128, 0)               = 0
futex(0x1722198, FUTEX_WAIT, 0, NULL)   = 0
epoll_wait(4, {}, 128, 0)               = 0
futex(0x1721420, FUTEX_WAKE, 1)         = 1
futex(0x17213a0, FUTEX_WAKE, 1)         = 1
epoll_wait(4, {}, 128, 0)               = 0
futex(0x1722198, FUTEX_WAIT, 0, NULL
```
However, if I `sudo service docker stop` and then execute it directly with `sudo unshare -m -- /usr/bin/docker daemon`, I get the following on the console:
```
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
INFO[0000] [graphdriver] using prior storage driver "devicemapper" 
INFO[0000] Option DefaultDriver: bridge                 
INFO[0000] Option DefaultNetwork: bridge                
INFO[0000] Firewalld running: true                      
INFO[0000] Loading containers: start.                   

INFO[0000] Loading containers: done.                    
INFO[0000] Daemon has completed initialization          
INFO[0000] Docker daemon                                 commit=f4bf5c7 execdriver=native-0.2 graphdriver=devicemapper version=1.8.3
```
And executing `sudo docker version` returns:
```
$ sudo docker version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 06:20:54 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 06:20:54 UTC 2015
 OS/Arch:      linux/amd64
```
with
```
INFO[0063] GET /v1.20/version
```
in the server terminal.

### Expected Results
The client should report the version and other information when the docker daemon is started via `sudo service docker start` and not just when it is started manually.

**Comment by phemmer** [comment]
Your commands show you mixing systemd, sysvinit, and manual commands. This is a likely the cause of problems. When you launch docker manually, it clobbers the socket file used by systemd.

Also as a side note, Fedora 20 is EOL, and not supported: https://docs.docker.com/engine/installation/fedora/

**Comment by csams** [comment]
I'm not mixing the start commands.  I'm demonstrating that the hang occurs when the docker daemon is started any way but manually.  If I start it manually, it works fine.  This is true even after killing the daemon started by systemd or sysvinit, blowing away the sock and pid files if they exist, ensuring the lock file is gone, and rebooting.  Starting via sysvinit OR systemd always causes the client to hang.

**Comment by cpuguy83** [comment]
@csams Can you paste the log output for when it doesn't work?
If upstart it's /var/log/upstart/docker, if systemd use journalctl.
I suspect systemd is timing out and killing it, and thus causing the client to hang until the client socket request times out.

**Comment by csams** [comment]
@cpuguy83 The request never times out - both the client and daemon just sit there.
```
[27078 csams@localhost ~]$ sudo systemctl status docker
docker.service - LSB: start and stop docker
   Loaded: loaded (/etc/rc.d/init.d/docker)
   Active: active (running) since Mon 2016-01-11 11:18:05 CST; 2min 10s ago
  Process: 23887 ExecStop=/etc/rc.d/init.d/docker stop (code=exited, status=0/SUCCESS)
  Process: 23899 ExecStart=/etc/rc.d/init.d/docker start (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/docker.service
           └─23910 /usr/bin/docker daemon

Jan 11 11:18:04 localhost.localdomain docker[23899]: Redirecting to /bin/systemctl status  cgconfig.service
Jan 11 11:18:05 localhost.localdomain docker[23899]: Starting docker:        [  OK  ]
Jan 11 11:18:05 localhost.localdomain systemd[1]: Started LSB: start and stop docker.

[27079 csams@localhost ~]$ ps -ef | grep docker
root     23910     1  0 11:18 ?        00:00:00 /usr/bin/docker daemon
csams    23985  3756  0 11:20 pts/0    00:00:00 grep --color docker

[27080 csams@localhost ~]$ tail /var/log/docker 
\nMon Jan 11 11:18:04 CST 2016\n
time="2016-01-11T11:18:04.278981704-06:00" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)" 
time="2016-01-11T11:18:04.351377001-06:00" level=info msg="[graphdriver] using prior storage driver \"devicemapper\"" 
time="2016-01-11T11:18:04.352269485-06:00" level=info msg="Option DefaultDriver: bridge" 
time="2016-01-11T11:18:04.352283233-06:00" level=info msg="Option DefaultNetwork: bridge" 

[27082 csams@localhost ~]$ sudo journalctl -u docker.service
Jan 11 11:18:04 localhost.localdomain systemd[1]: Starting LSB: start and stop docker...
Jan 11 11:18:04 localhost.localdomain docker[23899]: Redirecting to /bin/systemctl status  cgconfig.service
Jan 11 11:18:05 localhost.localdomain docker[23899]: Starting docker:        [  OK  ]
Jan 11 11:18:05 localhost.localdomain systemd[1]: Started LSB: start and stop docker.
```
Nothing in the status or log files change when I attempt to hit the daemon with `sudo docker version` or anything else.  The client just  hangs.  However, if I `sudo systemctl stop docker` and modify `/etc/init.d/docker` to include strace for the daemon before restarting it again
```
       "$unshare" -m -- strace -s 200 $exec daemon $other_args &>> $logfile &
#      "$unshare" -m -- $exec daemon $other_args &>> $logfile &
```
I get the following in `/var/log/docker`:
```
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=3439871008, u64=140075208295456}}}, 128, -1) = 1
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=3439871008, u64=140075208295456}}}, 128, -1) = 1
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=3439871008, u64=140075208295456}}}, 128, -1) = 1
epoll_wait(4, 
```
Each epoll_wait(...) represents an attempt to execute `sudo docker version`.

Now, if I do this:
```
[27091 csams@localhost ~]$ sudo systemctl stop docker
[27092 csams@localhost ~]$ sudo unshare -m -- /usr/bin/docker daemon &> docker.out &
[1] 24194
[27093 csams@localhost ~]$ tail -f docker.out 
time="2016-01-11T11:30:46.676165950-06:00" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)" 
time="2016-01-11T11:30:46.746697832-06:00" level=info msg="[graphdriver] using prior storage driver \"devicemapper\"" 
time="2016-01-11T11:30:46.747462275-06:00" level=info msg="Option DefaultDriver: bridge" 
time="2016-01-11T11:30:46.747475469-06:00" level=info msg="Option DefaultNetwork: bridge" 
time="2016-01-11T11:30:46.753364718-06:00" level=info msg="Firewalld running: true" 
time="2016-01-11T11:30:46.856794955-06:00" level=info msg="Loading containers: start." 

time="2016-01-11T11:30:46.857284824-06:00" level=info msg="Loading containers: done." 
time="2016-01-11T11:30:46.857305444-06:00" level=info msg="Daemon has completed initialization" 
time="2016-01-11T11:30:46.857318489-06:00" level=info msg="Docker daemon" commit=f4bf5c7 execdriver=native-0.2 graphdriver=devicemapper version=1.8.3 
```
and then in another terminal execute `sudo docker version`, I get
```
time="2016-01-11T11:31:07.685623613-06:00" level=info msg="GET /v1.20/version" 
```
logged from the daemon and
```
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 06:20:54 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 06:20:54 UTC 2015
 OS/Arch:      linux/amd64
```
back from the client.

One obvious difference when I execute the daemon manually is that I see
```
time="2016-01-11T11:30:46.753364718-06:00" level=info msg="Firewalld running: true" 
time="2016-01-11T11:30:46.856794955-06:00" level=info msg="Loading containers: start." 

time="2016-01-11T11:30:46.857284824-06:00" level=info msg="Loading containers: done." 
time="2016-01-11T11:30:46.857305444-06:00" level=info msg="Daemon has completed initialization" 
time="2016-01-11T11:30:46.857318489-06:00" level=info msg="Docker daemon" commit=f4bf5c7 execdriver=native-0.2 graphdriver=devicemapper version=1.8.3
```
after the Option DefaultDriver and Option DefaultNetwork lines, and that never shows up when starting with systemctl.

**Comment by unclejack** [comment]
@csams Can you reproduce this with the latest release of Docker?

**Comment by alexanderilyin** [comment]
@csams check docker logs (in my case it was in `/var/log/messages`) and if there http problems with 408 HTTP Status problem could be in current `mtu` value.

I had similar problem right now. Docker hangs during image pull. `mtu` was `9000`, workaround in my case:

```
sudo /sbin/ifconfig eth0 mtu 1500 up
```

**Comment by thaJeztah** [comment]
I'll close this issue, because this is for an old version of docker and may have been resolved; but please comment if you're still seeing this after upgrading to the current version of docker

**Comment by HuemanF** [comment]
Only manually running dameon works for me, hope this dump helps.
I still have this issue on 

4.8.13-100.fc23.x86_64

docker version
Client:
 Version:      1.12.5
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   7392c3b
 Built:        Fri Dec 16 02:26:11 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.5
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   7392c3b
 Built:        Fri Dec 16 02:26:11 2016
 OS/Arch:      linux/amd64



docker info
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 4
Server Version: 1.12.5
Storage Driver: devicemapper
 Pool Name: docker-253:1-1337297-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 2.107 GB
 Data Space Total: 107.4 GB
 Data Space Available: 24.79 GB
 Metadata Space Used: 2.445 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.109 (2015-09-22)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.8.13-100.fc23.x86_64
Operating System: Fedora 23 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 11.66 GiB
Name: localhost.localdomain
ID: R7QO:CX6H:6M4S:5PQ5:JZ57:ERET:WM3G:6E7Y:QUM2:LRTL:IE6K:XTAT
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8

**Comment by thaJeztah** [comment]
@HuemanF what happens if you start the service (`sudo systemctl start docker`) does it not start at all? do logs show anything?
